### DISCUSSION TITLE
Difficulty

### DISCUSSION THREAD

**Discussion by marcelharing** [beat-feet/beat-feet#78] [Ideas]
Great game, however it's still quite difficult (at least for me, but I am not a very skillfull gamer). Therefore maybe could you implement various levels of difficulty? For example a "beginner" and an "expert" level with the difference that as beginner you have more and faster regenerated shield points.

**Comment by vi** [comment]
Aren't insta-K.O. obstacles that are game over even with full hearts and full shields a bit overmuch?

Also should getting hit maybe refill the double-jump ability? "non-working" second jumps is what typically what causes the hits in the first place and that should prevent chains of getting hit again and again.

**Reply by pserwylo** [reply]
Could you let me know which level (and approx which % through) these obstacles are? The issue #67 is to track the fact that there are sometimes obstacles hidden behind other obstacles, which have no need to exist and just make it feel like you are taking mor damage. My hunch is that fixing that issue will address "insta-KO obstacles", but if not, we will need to do something else to limit the max damage an obstacle can take.

Also keen to understand more about "non-working" second jumps if you could elaborate a bit further. The intent, as you've probably noticed, is that they are designed to work *near* (but not exactly at) the top of your first jump. If you get hit on the way up, you can still double jump at the top of your jump. However if you are already on your way down, yo ucan't double jump again regardless of whether you get hit or not. This is inspired by the way double jumping worked in the original Unreal Tournament and other FPS games.

**Reply by vi** [reply]
Doing the second jump on the way down does seem to work in Beat Feet and is useful to prevent "impending disaster" or for covering large gaps.

However, if player sprite lands on a corner, a bit too early, it continues to "fall through" the ground and second jump (already depleted before) is not recharged, preventing the quick escape from the further damage. And getting hit again and again seems to prevent double jumps entirely, trapping the player at the bottom of some pillar series. 

> This is inspired by the way double jumping worked in the original Unreal Tournament and other FPS games.

For me it seems close to "I Wanna By The Guy: The Moive: The Game" mechanics. A "non-working" second jump (because of the first jump was actually falling off a platform instead of a proper jump) is a nuisance in that game as well.

**Reply by pserwylo** [reply]
Thanks for the feedback. If you like, I can try to come up with a couple of variations if you would like to test some?

**Reply by vi** [reply]
Yes, I can test them.

**Reply by vi** [reply]
> The intent, as you've probably noticed, is that they are designed to work near (but not exactly at) the top of your first jump. ...  However if you are already on your way down, you can't double jump ...

Tested it specifically and indeed the double jump does not work on the way down - I haven't noticed it before and assumed it just sometimes works, sometimes not.

Shall there be a tutorial mode that shows detailed rules about how specifically double jump works and how hits are counted and how hits and double jumps interact?

**Comment by comradekingu** [comment]
I like that it is a bit difficult. I could only clear a few levels before the glitching.
There is no real incentive to keep playing past getting all the levels, and trying many times to do it seems like a boring way to get there.
In Wario on gameboy, Wario collects various items because he knows he isn't as handsome as Mario, so it was fun to collect all the artifacts there. Not sure what the lore is in this game. :)

**Comment by pserwylo** [comment]
I think I'm in agreance with @comradekingu about the difficulty. There is a little bit of a discussion around the topic of difficulty and how difficult or easy it should be buried in [this issue](https://github.com/beat-feet/beat-feet/issues/63#issuecomment-791188981). 

I must admit that I've lost track of how easy or difficult it is, because the countless times I've play tested it has made it all pretty natural to me, to the point where I can start with a fresh install, and play each level once, unlocking enough achievements with each play through that I can almost unlock all the levels without having to replay a level again. However, I don't think the author of the game is a good metric for whether the difficulty is okay for new users or not.

Therefore, I think the notion of an "Easy", "Normal", "Hard" setting would be good. Initially this could just change the amount of damage you take, but then later it could even change the layout of the level a little (though that is much more long term).

**Comment by vi** [comment]
Easy mode idea: jumping in mid-air always to work if player sprite is directly above any lamp post (but such jumps do not count towards a combo).

**Comment by paulusasol** [comment]
I personally think the difficulty overall is fine (but of course, there is still potential for optimization, as mentioned in other comments, but I think there is some lack of low-difficulty levels.

#### My experience getting through the levels

1. I play the first few levels, playing the levels multiple times to progress further.
2. I finish the first level (reach 100% in it).
3. I continue playing until I unlock all levels, adding a custom song during this time.
4. I keep playing to finish all levels and swiftly achieve it in the custom song<sup>[1](#custom-song-footnote)</sup> & Vivaldi.
5. I do not finish any further level for a significant amount of time.
6. I finish six further levels in a very short amount of time.

#### What I would expect:
There are no large gaps between finishing the first two to three levels and finishing the third to fourth level. Also I think finishing multiple early levels should be something to (normally) be achieved before unlocking all levels.

<a name="custom-song-footnote">1</a>: Actually less than 40% due to issue #85.

**Reply by vi** [reply]
Similar experience. First level got 100% almost immediately, second level took much longer. Third one also took much longer, but after that most of levels got completed one after another. "Contact with Dust and Lint" proved to be somehow more problematic, and the antenna is still unfinished (it's too long).

**Reply by pserwylo** [reply]
Thanks @paulusasol for the detailed summary (likewise @vi). Although I'm kind of aware there are people playing and (hopefully?) enjoying the game, I have no idea about whether it is *good* or easy/hard it is. Feedback like this is vital. During developmnt I've had to play test so many times that I'm no longer a good measure of whether it is hard, because I kind of know all the kinks and difficult bits in each level, and I have developed a good feel for how the player jumps.

I agree that specific achievements should be introduced, so that instead of just "5 achievements" you have "2 completed levels and 3 at 50%" or something like that.

Can I ask your opinion on whether we should re-jig the game to be more like a Super Mario Brothers World? 

Something like this:

![Untitled](https://user-images.githubusercontent.com/248565/125159969-e612ea80-e1bd-11eb-8cbe-778706c554ba.png)

If so, what would be the criteria for getting to the next level? Given quite a few people find it quite difficult to clear levels, I don't think we could have a Mario-style "Finish this level to go to the next". Perhaps placing barriers on the road *between* levels, whereby you need to meet certain achievements (as discussed above) in order to clear the barrier...

**Reply by comradekingu** [reply]
Maybe the hitbox of objects could be reduced the tiniest bit. After it changed to put the player on top in case of doubt, sometimes I clear when I feel I shouldn't.
I got 100% on the first 9 levels, but then 16-61% on the rest.
A screen that you move through by jumping to select the level seems good.
One loop with 4 levels in each.
Build a sturdier road by trying 25 times, get a better symbol for the level by clearing %, and don't know what for combo.
Same jump-style selector could be done with the menu after a game.
Run the player underneath the options scrolling beneath.

**Reply by vi** [reply]
> criteria for getting to the next level?

Why shouldn't all levels should be accessible by default in the first place? Currently it is rather easy to unlock all the levels (as achievements are often granted for efforts, not for results).

At least it should not reduce to system where you need to 100%-complete level N to gain access to level N+1, so that one tricky level prevents progression to all the next levels.

---

There maybe two distinct modes, like "story" mode that contains predefined levels and counts progression, and custom level-oriented mode (but with predefined levels also available), which mostly tracks stats and treats all levels as equivalent.

**Reply by comradekingu** [reply]
Because otherwise there is no privileged reward and no challenge getting there.
There are enough levels to pick from, and the experienced variety increases by getting somewhere.

**Comment by ghost** [comment]
I too think there is a difficulty issue. The second level is a difficulty spike. Maybe add 1 or 2 levels between that?

**Reply by ghost** [reply]
I think Maintainance should be before The Courtyard.
